include/dali/core/mm/cuda_vm_resource.h

```diff
@@ -191,13 +205,124 @@ class cuda_vm_resource : public memory_resource<memory_kind::device>,
+    ranges_to_free.reserve(va_ranges_.size());
+    size_t va_freed = 0;
+
+    {
+      std::unique_lock pool_guard(pool_lock_, std::try_to_lock);
+      std::unique_lock mem_guard(mem_lock_, std::try_to_lock);
+
+      // the whole block below musn't throw or it'll leave the resource in an inconsistent state
+      [&]() noexcept {
+        for (int i = 0, n = va_ranges_.size(); i < n; i++) {
+          auto start = va_ranges_[i].ptr();
+          auto end = va_ranges_[i].end();
+          if (free_va_.contains(reinterpret_cast<void*>(start), reinterpret_cast<void*>(end))) {
+            int region_idx = find_va_region(start);
+            assert(region_idx >= 0);
+            va_region &region = va_regions_[region_idx];
+
+            ptrdiff_t offset = region.address_range.ptr() - start;
+            assert(offset % block_size_ == 0);
+            int num_blocks = (end - start) / block_size_;
+            int start_block = offset / block_size_;  // index of the first block
+            int end_block = start_block + num_blocks;  // index of the last block
+
+            int avail = region.available.find(true, start_block);
+            while (avail < end_block) {
+              int end_avail = std::min<int>(region.available.find(false, avail + 1), end_block);
+              for (int b = avail; b < end_avail; b++) {
+                blocks_to_free.push_back(region.unmap_block(b));
+                stat_.total_unmaps++;
+              }
+              auto *start_ptr = region.block_ptr<char>(avail);
+              auto *end_ptr = region.block_ptr<char>(end_avail);
+              free_mapped_.get_specific_block(start_ptr, end_ptr);
+
+              if (end_avail == end_block)
+                break;
+              avail = region.available.find(true, end_avail + 1);
+            }
+
+            ranges_to_free.push_back({ i, region_idx, start_block, end_block });
+          }
+        }
+        stat_.allocated_blocks -= blocks_to_free.size();
+      }();
+
+      for (const RangeDesc &rd : ranges_to_free) {
+        auto &region = va_regions_[rd.region_idx];
+
+        // We split the va_region as follows:
+
+        // |<--------------- old region ---------------------->|
+        // |<---region-->|<--va_range_to_remove-->|<---tail--->|
+        //               ^                        ^
+        // start_block---^                        ^---- end_block
+
+        // If start_block is 0 then `region` will become empty and we can either
+        // overwrite it with tail (if not empty) or remove it altogether
+
+        va_region tail = region.split(rd.end_block);
+        region.resize(rd.start_block);
+
+        auto *start_ptr = region.block_ptr<char>(rd.start_block);
+        auto *end_ptr = region.block_ptr<char>(rd.end_block);
+        ptrdiff_t range_size = (end_ptr - start_ptr);
+
+        if (region.empty()) {
+          if (tail.empty())
+            va_regions_.erase(va_regions_.begin() + rd.region_idx);
```

Comment:
I don't fully understand that case - if both region and tail are empty does it mean that the old region was empty as well?

Reply:
It means that the range covered the entire region: the tail is empty and the head is shrunk to zero blocks.